### PR TITLE
Reworked Go func to del old projects

### PR DESCRIPTION
**Pull request by paulpalamarchuk**
* Fix https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/issues/400

### CODE REVIEW

File: modules/project_cleanup/function_source/main.go [hunk 1]
```diff
@@ -17,99 +17,192 @@ limitations under the License.
...
+	includedLabelsMap := getLabelsMapFromEnv(TargetIncludedLabels)
+	includeProjectByOneOfLabelsFilter := func(project *cloudresourcemanager.Project) bool {
+		return checkIfAtLeastOneLabelPresentIfAny(project, includedLabelsMap, false)
+	}
+
+	resourceCreationCutoff := tooOldTime(int64(getCorrectMaxAgeInHoursOrTerminateExecution()) * 60 * 60)
+	ageFilter := func(project *cloudresourcemanager.Project) bool {
+		projectCreatedAt, err := time.Parse(time.RFC3339, project.CreateTime)
+		if err != nil {
+			logger.Println(fmt.Sprintf("Fail to parse CreateTime for [%s], skip it. Error [%s]", project.Name, err.Error()))
```
Comment (kopachevsky) [inline modules/project_cleanup/function_source/main.go]: use logger.Printf for formatting

**Comment by morgante** [comment]
@paulpalamarchuk Is this ready for review? We need to get this done ASAP.

File: modules/project_cleanup/function_source/main.go [hunk 2]
```diff
@@ -17,99 +17,192 @@ limitations under the License.
...
-// CleanUpProjects is the entry point of the scheduled function.
-func CleanUpProjects(ctx context.Context, m PubSubMessage) error {
-	targetTag := os.Getenv("TARGET_TAG_NAME")
-	targetValue := os.Getenv("TARGET_TAG_VALUE")
-	maxAgeInHoursStr := os.Getenv("MAX_PROJECT_AGE_HOURS")
-	maxAgeInHours, err := strconv.ParseInt(maxAgeInHoursStr, 10, 0)
-	if err != nil {
-		logger.Fatal(fmt.Sprintf("Could not convert %s to integer", maxAgeInHoursStr))
-		return err
+const LifecycleStateActiveRequested = "ACTIVE"
```
[the remark this answers is not in the document]
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: Moved

File: modules/project_cleanup/function_source/main.go (hunk 1 above)
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: updated

File: modules/project_cleanup/function_source/main.go [hunk 3]
```diff
@@ -17,99 +17,237 @@ limitations under the License.
...
 	if err != nil {
-		return err
+		logger.Println("Fail to initialize Google client, terminate execution")
+		logger.Fatal(err)
 	}
+	logger.Println("Initialized Google client")
+	logger.Println("Try to get Cloud Resource Manager")
+	cloudResourceManagerService, err := cloudresourcemanager.New(client)
+	if err != nil {
+		logger.Println("Fail to get Cloud Resource Manager, terminate execution")
```
Comment (kopachevsky) [inline modules/project_cleanup/function_source/main.go]: use https://golang.org/pkg/log/#Fatalf instead 2 lines

File: modules/project_cleanup/function_source/main.go [hunk 4]
```diff
@@ -17,99 +17,237 @@ limitations under the License.
...
+	return cloudResourceManagerService
+}
 
-	resourceCreationCutoff := tooOldTime(int64(acceptableAgeInHours * 60 * 60))
+func getFolderServiceOrTerminateExecution(ctx context.Context) *cloudresourcemanager2.FoldersService {
+	logger.Println("Try to initialize Google client")
+	client, err := google.DefaultClient(ctx, cloudresourcemanager.CloudPlatformScope)
+	if err != nil {
+		logger.Println("Fail to initialize Google client, terminate execution")
+		logger.Fatal(err)
```
Comment (kopachevsky) [inline modules/project_cleanup/function_source/main.go]: same

File: modules/project_cleanup/function_source/main.go [hunk 5]
```diff
@@ -17,99 +17,237 @@ limitations under the License.
...
+	}
+	logger.Println("Initialized Google client")
+	logger.Println("Try to get Folders Service")
+	cloudResourceManagerService, err := cloudresourcemanager2.New(client)
+	if err != nil {
+		logger.Println("Fail to get Folders Service, terminate execution")
+		logger.Fatal(err)
```
Comment (kopachevsky) [inline modules/project_cleanup/function_source/main.go]: same

File: modules/project_cleanup/function_source/main.go [hunk 6]
```diff
@@ -17,99 +17,237 @@ limitations under the License.
...
-		for _, project := range page.Projects {
-			if val, ok := project.Labels[key]; !(ok && val == value) {
-				logger.Println(fmt.Sprintf("Rejecting project %s because of missing label", project.ProjectId))
-				continue
-			}
+func invoke(ctx context.Context) {
+	cloudResourceManagerService := getResourceManagerServiceOrTerminateExecution(ctx)
+	folderService := getFolderServiceOrTerminateExecution(ctx)
+
+	removeLien := func(name string) {
```
Comment (kopachevsky) [inline modules/project_cleanup/function_source/main.go]: what is lien?

File: modules/project_cleanup/function_source/main.go [hunk 7]
```diff
@@ -17,99 +17,237 @@ limitations under the License.
...
+		listFoldersRequest := folderService.List().Parent(folderId).ShowDeleted(false)
+		if err := listFoldersRequest.Pages(ctx, func(foldersResponse *cloudresourcemanager2.ListFoldersResponse) error {
+			for _, folder := range foldersResponse.Folders {
+				recursion(folder.Name, recursion)
 			}
-			logger.Println(fmt.Sprintf("Requested deletion of project %s.", project.ProjectId))
+			removeProjectsInFolder(folderId)
+			return nil
+		}); err != nil {
+			logger.Fatal(fmt.Sprintf("Fail to get subfolders for the forlder with id [%s], error [%s]", folderId, err.Error()))
```
Comment (kopachevsky) [inline modules/project_cleanup/function_source/main.go]: https://golang.org/pkg/log/#Fatalf

File: modules/project_cleanup/function_source/main.go (hunk 6 above)
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: https://cloud.google.com/resource-manager/docs/project-liens

File: modules/project_cleanup/function_source/main.go (hunk 3 above)
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: updated

File: modules/project_cleanup/function_source/main.go (hunk 4 above)
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: updated

File: modules/project_cleanup/function_source/main.go (hunk 5 above)
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: updated

File: modules/project_cleanup/function_source/main.go (hunk 7 above)
Comment (paulpalamarchuk) [reply modules/project_cleanup/function_source/main.go]: updated

**Comment by paulpalamarchuk** [comment]
@morgante PR ready for review

File: modules/project_cleanup/variables.tf [hunk 8]
```diff
@@ -29,20 +29,50 @@ variable "region" {
...
 
 variable "max_project_age_in_hours" {
   type        = number
   description = "The maximum number of hours that a GCP project, selected by `target_tag_name` and `target_tag_value`, can exist"
   default     = 6
 }
+
+variable "target_excluded_labels" {
+  type        = map(string)
+  description = "Map of project lablels that won't be deleted."
```
Comment (morgante) [inline modules/project_cleanup/variables.tf]:
```suggestion
  description = "Map of project labels that won't be deleted."
```

File: modules/project_cleanup/variables.tf [hunk 9]
```diff
@@ -29,20 +29,50 @@ variable "region" {
...
+  default     = {}
+}
+
+variable "target_included_labels" {
+  type        = map(string)
+  description = "Map of project lablels that will be deleted."
```
Comment (morgante) [inline modules/project_cleanup/variables.tf]:
```suggestion
  description = "Map of project labels that will be deleted."
```

File: modules/project_cleanup/function_source/go.sum [hunk 10]
```diff
@@ -1,115 +0,0 @@
-cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (morgante) [inline modules/project_cleanup/function_source/go.sum]: Please don't remove `go.sum`.

**Pull request merged**